text/0000-cloud-based-mapping.md

```diff
@@ -0,0 +1,134 @@
+
+- *Cloud Cartographer* does not require a `LocalTrajectoryBuilder` as there is no local SLAM problem to be solved.
+- *Agent Cartographer* needs to upload all necessary information for *Cloud Cartographer* to formulate the global SLAM problem *and* also route that information to its local version of the global optimization problem that is reduced to the pure localization case.
+
+### Software Components Overview
+
+While implementing this RFC we want to keep the following three types of Cartographer users happy
+
+- Users that link against libcartographer as a C++ library and don’t care about ROS at all
+- ROS users that use the ROS binaries *cartographer_offline_node* and *cartographer_node* to stream sensor data as ROS messages and receive Cartographer’s SLAM results as ROS messages
+- Users that want to use cloud-based mapping, might use ROS or not to send sensor data into the system and are happy to embrace gRPC as an IPC system both on the robot and between robot and cloud.
+
+To make sure all three groups of users are comfortable with the plans laid out in this RFC we have decided to
+
+- continue our efforts to push for a stable public libcartographer C++ interface.
+- keep *cartographer_offline_node* and *cartographer_node* free of gRPC serialization/deserialization.
+- introduce new gRPC-dependent binaries in the cartographer and cartographer_ros repo that form the basis of cloud-based mapping.
+- hide any dependency on gRPC behind a compilation flag.
+
+#### cartographer_node_grpc and cartographer_offline_node_grpc
+
+In the *cartographer_ros* repository we will add two new binaries, *cartographer_node_grpc* and *cartographer_offline_node_grpc*.
+These binaries will contain the same ROS specific code (e.g. `Node`) that *cartographer_node* and *cartographer_offline_node* contain, e.g. they take care of all the TF logic, but perform gRPC calls against the *cartographer_grpc* binary (see below) for everything that was formerly passed through C++ function invocations against `MapBuilderBridge` and `SensorBridge`.
+
+This will require us to
+
+- extract interfaces out of `MapBuilderBridge` and `SensorBridge`.
+- implement a gRPC bridging version of the above two classes that serialize any sensor data to proto3 and send that to a gRPC service.
+
+In principle once we have a binary that does ROS to gRPC translation it would be possible to replace the existing cartographer ROS node with a combination of this translation binary and *cartographer_grpc*, i.e. instead of the C++ interface between *cartographer_node* and libcartographer that exists today we would introduce a local, loopback gRPC layer.
+We have decided not to go this route and keep the current *cartographer_node* as is until we have a better idea of the performance impact (or benefit) of this additional gRPC serialization / deserialization and whether adding gRPC to our dependencies is acceptable for our ROS users.
+
+#### cartographer_grpc
```

Comment:
This is quite hand wavy. Will there be a follow RFC that explains that thing in more detail?